packages/server/lib/crons/deleteOldData.ts

```diff
@@ -50,62 +70,226 @@ export async function exec(): Promise<void> {
+                }
+
+                return integrations.length;
+            }
+        });
+
+        // Delete sync configs and all associated data
+        await batchDelete({
+            name: 'sync configs',
+            startTimestamp,
+            ttlMs,
+            deleteFn: async () => {
+                const syncsConfigs = await getSoftDeletedSyncConfig({ limit, olderThan: deleteSyncConfigsOlderThan });
+
+                for (const syncConfig of syncsConfigs) {
+                    await deleteSyncConfigData({ syncConfig, startTimestamp, ttlMs });
+                }
+
+                return syncsConfigs.length;
             }
-            if (Date.now() - startTimestamp > ttlMs) {
-                logger.info(`Time limit reached, stopping`);
-                return;
+        });
+
+        // Delete connections and all associated data
+        await batchDelete({
+            name: 'connections',
+            startTimestamp,
+            ttlMs,
+            deleteFn: async () => {
+                const connections = await connectionService.getSoftDeleted({ limit, olderThan: deleteConnectionsOlderThan });
+
+                for (const connection of connections) {
+                    logger.info('Deleting connection...', connection.id, connection.connection_id);
+                    const resSyncs = await db.knex
+                        .select<
+                            {
+                                sync: Sync;
+                                syncConfig: DBSyncConfig;
+                            }[]
+                        >(db.knex.raw('row_to_json(_nango_syncs.*) as sync'), db.knex.raw('row_to_json(_nango_sync_configs.*) as "syncConfig"'))
+                        .from<Sync>('_nango_syncs')
+                        .join('_nango_sync_configs', '_nango_sync_configs.id', '_nango_syncs.sync_config_id')
+                        .where({ nango_connection_id: connection.id });
+                    for (const res of resSyncs) {
+                        await deleteSyncData({ ...res });
+                    }
+
+                    await connectionService.hardDelete(connection.id);
+                }
+
+                return connections.length;
             }
-            await setTimeout(1000);
-        }
+        });
     } finally {
         if (lock) {
             locking.release(lock);
         }
     }
 }
+
+async function batchDelete({
+    name,
+    deleteFn,
+    startTimestamp,
+    ttlMs
+}: {
+    name: string;
+    deleteFn: () => Promise<number>;
+    startTimestamp: number;
+    ttlMs: number;
+}) {
+    while (true) {
+        const deleted = await deleteFn();
+        if (deleted) {
+            logger.info(`Deleted ${deleted} ${name}`);
+        }
+        if (deleted < limit) {
+            break;
+        }
+        if (Date.now() - startTimestamp > ttlMs) {
+            logger.info(`Time limit reached, stopping`);
+            return;
+        }
+        await setTimeout(1000);
+    }
+}
```

Comment:
should `deleteFn` be passed a transaction? If the function executes multiples queries and one fails, we would end up with a partial deletion

Reply:
the problem with transaction is that it needs to keep everything in memory which is not always possible, degrading the whole instance performance.
If one fail, it will throw, bubble up and stops the cron starting back from where it was next time